eip-4626.md

```diff
@@ -0,0 +1,152 @@
+---
+eip: 4626
+title: Yield Bearing Vault Standard
+description: A standard for yield bearing vaults.
+author: Joey Santoro (@joeysantoro), t11s (@transmissions11), Jet Jadeja (@JetJadeja)
+discussions-to: https://ethereum-magicians.org/t/eip-4626-yield-bearing-vault-standard/7900
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-12-22
+---
+
+## Abstract
+
+The following standard allows for the implementation of a standard API for yield bearing vaults within smart contracts. This standard provides basic functionality for depositing and withdrawing tokens and reading balances with an optional extension for tokenized vaults using ERC-20.
+
+## Motivation
+
+Yield bearing vaults have a lack of standardization leading to diverse implementation details. Some various examples include lending markets (Compound, Aave, Fuse), aggregators (Yearn, Rari Vaults, Idle), and intrinsically interest bearing tokens (xSushi). This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards. This forces each protocol to implement their own adapters which are error prone and waste development resources.
+
+A standard for yield bearing vaults will allow for a similar cambrian explosion to ERC-20, unlocking access to yield in a variety of applications with little specialized effort from developers.
+
+
+## Specification
+
+### Methods
+
+#### deposit
+
+`function deposit(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Deposits `_value` tokens into the vault and grants ownership of them to `_to`.
+
+MAY return a pro-rata ownership `_shares` value corresponding `_value`, if not MUST return `0`.
+
+MAY represent `_shares` using internal accounting or an ERC-20 token. 
+
+If pro-rata shares ownership is implemented, the vault SHOULD implement `balanceOf`, `redeem`, `totalSupply` and `exchangeRate`.
+
+
+#### withdraw
+
+`function withdraw(address _to, uint256 _value) public returns (uint256 _shares)`
```

Comment:
I am not as opinionated on the naming conventions as @transmissions11 and am open to reordering or making `redeem` required which is the shares based model. I do like the symmetry of deposit/withdraw both using the same underlying token accounting and having redeem or withdrawShares use the shares model. I'm also comfortable having withdraw be the shares and withdrawUnderlying be for underlying. Combined with @alcueca's feedback, I believe the standard can be somewhat more opinionated on some of these details.

Reply:
Agree on this, I think the norm of most protocols is to use `value` for depositing & `shares` for withdrawing, so making the default `withdraw` function to use `shares` will minimise unexpected surprises to devs. 
Since the interface is for a vault, having an additional `withdrawValue` is maybe necessary? But I think the devs can easily do some math and use the core `withdraw` function if necessary. This is probably better than forcing the function in the interface.

Reply:
Please, see above. I think that both the `withdraw` and `deposit` features should accept both `value` or `shares` as inputs.

Reply:
Happy to agree to disagree, I'm not making any statement about what I think the code should look like.

The code that is being used in production by the majority of protocols looks like the following
`function deposit(uint256 amount)`
`function withdraw(uint256 shares)`

Happy to go into detail as to why that is the case, but this interface has survived over hundreds of forks on all EVM compatible chains, and as such shows wisdom in its simplicity

Reply:
Also, withdrawing a fixed amount of underlying (rather than shares) from vaults always ends up leaving some dust behind

Reply:
hence why we have a redeem

Reply:
withdraw standing for shares is flat out illogical. you don't "withdraw" a check at the bank, or "withdraw" a winning lottery ticket.

Reply:
All of the cToken-based protocols (Compound, Rari, Cream, etc) use `mint(uint amount)`, `redeem(uint shares)`, `redeemUnderlying(uint amount)` so I think whichever way is chosen there's going to be a group of existing protocols that aren't natively compatible.

Reply:
> withdraw standing for shares is flat out illogical. you don't "withdraw" a check at the bank, or "withdraw" a winning lottery ticket. you **redeem** them. you withdraw $100

You can't withdraw to someone else either, the bank analogy falls flat

Reply:
> withdraw standing for shares is flat out illogical. you don't "withdraw" a check at the bank, or "withdraw" a winning lottery ticket. you **redeem** them. you withdraw $100

As for withdrawing $100 dollars, the reason why fork over fork never killed the `withdraw(shares)` is the level of complexity or close to impossibility to withdraw all shares.

Illustrated by literally the first withdraw I found on Rari's smart contract

https://etherscan.io/tx/0xcada09b35d35aaee76fdaf2a00b8f275b07ced7eb8c80cf48cc5dac42c4832ea

<img width="1141" alt="Screenshot 2022-01-07 at 15 40 02" src="https://user-images.githubusercontent.com/13383782/148559659-7262784e-ec04-47d3-af6e-802bc6f5f293.png">

<img width="486" alt="Screenshot 2022-01-07 at 15 40 14" src="https://user-images.githubusercontent.com/13383782/148559687-d6ab614d-3205-4cc1-a927-024ee94c809b.png">

Did the depositor really want to keep those 0.02 tokens? Or is the implementation that forces them into something they don't want?

Reply:
@GalloDaSballo

**Definitions**
(Google) withdraw: remove or take away (something) from a particular place or position
(Merriam Webster) withdraw: to remove (money) from a place of deposit

It doesn't seem to say anything about where the money ends up (the destination), so I think you **can** withdraw into *someone else's* account

Reply:
> the reason why fork over fork never killed the `withdraw(shares)` is the level of complexity or close to impossibility to withdraw all shares

Regardless, the term "redeem" is perhaps better-fitted than "withdraw".